engine/src/main/java/io/zeebe/engine/processor/workflow/message/BufferedMessageToStartEventCorrelator.java

```diff
@@ -0,0 +1,177 @@
+
+  public BufferedMessageToStartEventCorrelator(
+      final KeyGenerator keyGenerator, final MessageState messageState) {
+    this.keyGenerator = keyGenerator;
+    this.messageState = messageState;
+  }
+
+  @Override
+  public void accept(final BpmnStepContext<ExecutableFlowElementContainer> context) {
+
+    final var workflowInstanceKey = context.getValue().getWorkflowInstanceKey();
+    final var correlationKey = messageState.getWorkflowInstanceCorrelationKey(workflowInstanceKey);
+
+    if (correlationKey != null) {
+      messageState.removeWorkflowInstanceCorrelationKey(workflowInstanceKey);
+
+      // the workflow instance was created by a message with a correlation key
+      // - other messages with same correlation key are not correlated to this workflow until this
+      // instance is ended (workflow-correlation-key lock)
+      // - now, after the instance is ended, correlate the next buffered message
+      correlateNextBufferedMessage(correlationKey, context);
+    }
+  }
+
+  private void correlateNextBufferedMessage(
+      final DirectBuffer correlationKey,
+      final BpmnStepContext<ExecutableFlowElementContainer> context) {
+
+    final var bpmnProcessId = context.getValue().getBpmnProcessIdBuffer();
+    final var workflow = context.getStateDb().getLatestWorkflowVersionByProcessId(bpmnProcessId);
+
+    final var messageCorrelation = findNextMessageToCorrelate(workflow, correlationKey);
+
+    if (messageCorrelation == null) {
+      // no buffered message to correlate
+      // - release the workflow-correlation-key lock
+      messageState.removeActiveWorkflowInstance(bpmnProcessId, correlationKey);
+
+    } else {
+      final var message = messageState.getMessage(messageCorrelation.messageKey);
+
+      correlateMessage(workflow, messageCorrelation.elementId, message, context);
+    }
+  }
+
+  private Correlation findNextMessageToCorrelate(
+      final DeployedWorkflow workflow, final DirectBuffer correlationKey) {
+
+    messageCorrelation.messageKey = Long.MAX_VALUE;
+    messageCorrelation.elementId = null;
+
+    for (final ExecutableStartEvent startEvent : workflow.getWorkflow().getStartEvents()) {
+      if (startEvent.isMessage()) {
+
+        final var messageName = startEvent.getMessage().getMessageName();
+
+        messageState.visitMessages(
+            messageName,
+            correlationKey,
+            message -> {
+              // correlate the first message with same correlation key that was not correlated yet
+              if (message.getDeadline() > ActorClock.currentTimeMillis()
+                  && !messageState.existMessageCorrelation(
+                      message.getKey(), workflow.getBpmnProcessId())) {
+
+                // correlate the first published message across all message start events
+                // - using the message key to decide which message was published before
+                if (message.getKey() < messageCorrelation.messageKey) {
```

Comment:
Assuming a workflow with start messages `a` and `b`. Since it iterates over the start messages, it looks for a message `a` first, then `b`. 

The expression is `false` if a message `b` was published after a message `a`.